File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/output/TableFormatter.java
package org.iumotionlabs.hefesto.feature.portinfo.output;

import org.iumotionlabs.hefesto.core.config.HefestoConfig;
import org.iumotionlabs.hefesto.core.port.output.OutputPort;
import org.iumotionlabs.hefesto.feature.portinfo.model.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Formats output as ASCII tables.
 */
public final class TableFormatter implements OutputFormatter {

    private final HefestoConfig config = HefestoConfig.get();
    private final boolean useColors;

    public TableFormatter() {
        this.useColors = true;
    }

    public TableFormatter(boolean useColors) {
        this.useColors = useColors;
    }

    @Override
    public String format(List<EnrichedPortBinding> bindings) {
        if (bindings.isEmpty()) {
            return "No port bindings found.";
        }

        // Calculate column widths (capped at configured max)
        int maxProcessWidth = config.maxProcessNameWidth();
        int maxServiceWidth = config.maxServiceWidth();

        int protoWidth = 5;
        int portWidth = 6;
        int stateWidth = 12;
        int addressWidth = 15;
        int pidWidth = 8;
        int processWidth = 15;
        int serviceWidth = 15;

        for (EnrichedPortBinding b : bindings) {
            int procLen = b.processName() != null ? b.processName().length() : 0;
            processWidth = Math.max(processWidth, maxProcessWidth > 0 ? Math.min(procLen, maxProcessWidth) : procLen);
            if (b.serviceInfo() != null) {
                int svcLen = b.serviceInfo().name().length();
                serviceWidth = Math.max(serviceWidth, maxServiceWidth > 0 ? Math.min(svcLen, maxServiceWidth) : svcLen);
            }
        }

        // Build header
        StringBuilder sb = new StringBuilder();
        String format = "%-" + protoWidth + "s  %-" + portWidth + "s  %-" + stateWidth + "s  %-" + addressWidth +
                       "s  %-" + pidWidth + "s  %-" + processWidth + "s  %-" + serviceWidth + "s%n";

        String header = String.format(format, "PROTO", "PORT", "STATE", "ADDRESS", "PID", "PROCESS", "SERVICE");
        String separator = "-".repeat(header.length() - 1);

        if (useColors) {
            sb.append(OutputPort.BOLD).append(header).append(OutputPort.RESET);
        } else {
            sb.append(header);
        }
        sb.append(separator).append("\n");

        // Build rows
        for (EnrichedPortBinding b : bindings) {
            String proto = b.protocol();
            String port = String.valueOf(b.port());
            String state = b.state();
            String address = b.localAddress();
            String pid = String.valueOf(b.pid());
            String process = config.truncateProcessName(b.processName());
            String service = b.serviceInfo() != null ? config.truncateService(b.serviceInfo().name()) : "-";

            if (useColors) {
                proto = colorize(proto, OutputPort.CYAN);
                state = colorizeState(state);
                process = colorize(process, OutputPort.GREEN);
                if (b.isExposed()) {
                    address = colorize(address, OutputPort.YELLOW);
                }
            }

            sb.append(String.format(format, proto, port, state, address, pid, process, service));
        }

        return sb.toString();
    }

    @Override
    public String formatOverview(NetworkOverview overview) {
        StringBuilder sb = new StringBuilder();

        // Title
        if (useColors) {
            sb.append(OutputPort.BOLD).append(OutputPort.UNDERLINE);
        }
        sb.append("NETWORK OVERVIEW\n");
        if (useColors) {
            sb.append(OutputPort.RESET);
        }
        sb.append("================\n\n");

        // Statistics
        sb.append("STATISTICS\n");
        NetworkOverview.Statistics stats = overview.statistics();
        sb.append("  Listening:     ").append(stats.totalListening()).append(" ports\n");
        sb.append("  Established:   ").append(stats.totalEstablished()).append(" connections\n");
        sb.append("  TCP: ").append(stats.tcpCount()).append(" | UDP: ").append(stats.udpCount()).append("\n");
        sb.append("  Exposed (0.0.0.0): ").append(stats.exposedCount());
        sb.append(" | Local (127.0.0.1): ").append(stats.localOnlyCount()).append("\n\n");

        // Exposed ports
        List<EnrichedPortBinding> exposed = overview.exposedPorts();
        if (!exposed.isEmpty()) {
            sb.append("EXPOSED PORTS (Network Accessible)\n");
            for (EnrichedPortBinding b : exposed) {
                String serviceTag = b.serviceInfo() != null ? b.serviceInfo().toTag() : "";
                String line = "  %s :%d   %-12s %-15s pid=%d%n".formatted(
                    b.protocol(), b.port(), b.processName(), serviceTag, b.pid()
                );
                if (useColors) {
                    sb.append(OutputPort.YELLOW);
                }
                sb.append(line);
                if (useColors) {
                    sb.append(OutputPort.RESET);
                }
            }
            sb.append("\n");
        }

        // Group by process
        Map<Long, List<EnrichedPortBinding>> byPid = overview.byPid();
        if (!byPid.isEmpty()) {
            sb.append("PROCESSES\n");
            for (Map.Entry<Long, List<EnrichedPortBinding>> entry : byPid.entrySet()) {
                List<EnrichedPortBinding> ports = entry.getValue();
                EnrichedPortBinding first = ports.get(0);

                if (useColors) {
                    sb.append(OutputPort.GREEN);
                }
                sb.append("  ").append(first.processName());
                sb.append(" (pid ").append(first.pid()).append(")");
                if (!first.user().isEmpty()) {
                    sb.append(" - ").append(first.user());
                }
                if (useColors) {
                    sb.append(OutputPort.RESET);
                }
                sb.append("\n");

                for (EnrichedPortBinding b : ports) {
                    String serviceTag = b.serviceInfo() != null ? b.serviceInfo().toTag() : "";
                    String exposure = b.isExposed() ? "0.0.0.0" : b.isLocalOnly() ? "127.0.0.1" : b.localAddress();
                    sb.append("    :").append(b.port());
                    sb.append("  ").append(b.state());
                    sb.append("  ").append(exposure);
                    sb.append("  ").append(serviceTag);
                    sb.append("\n");
                }
            }
        }

        return sb.toString();
    }

    @Override
    public String formatSecurityReport(SecurityReport report) {
        StringBuilder sb = new StringBuilder();

        // Title
        if (useColors) {
            sb.append(OutputPort.BOLD).append(OutputPort.UNDERLINE);
        }
        sb.append("SECURITY ANALYSIS\n");
        if (useColors) {
            sb.append(OutputPort.RESET);
        }
        sb.append("=================\n\n");

        if (report.isClean()) {
            if (useColors) {
                sb.append(OutputPort.GREEN);
            }
            sb.append("No security issues found.\n");
            if (useColors) {
                sb.append(OutputPort.RESET);
            }
            return sb.toString();
        }

        // Group by severity
        Map<SecurityFlag.Severity, List<SecurityFlag>> bySeverity = report.bySeverity();

        for (SecurityFlag.Severity severity : List.of(
            SecurityFlag.Severity.CRITICAL,
            SecurityFlag.Severity.HIGH,
            SecurityFlag.Severity.WARNING,
            SecurityFlag.Severity.INFO
        )) {
            List<SecurityFlag> flags = bySeverity.getOrDefault(severity, List.of());
            if (flags.isEmpty()) continue;

            String color = switch (severity) {
                case CRITICAL -> OutputPort.RED;
                case HIGH -> OutputPort.RED;
                case WARNING -> OutputPort.YELLOW;
                case INFO -> OutputPort.CYAN;
            };

            if (useColors) {
                sb.append(color).append(OutputPort.BOLD);
            }
            sb.append(severity.displayName().toUpperCase()).append(" (").append(flags.size()).append(")\n");
            if (useColors) {
                sb.append(OutputPort.RESET);
            }

            for (SecurityFlag flag : flags) {
                if (useColors) {
                    sb.append(color);
                }
                sb.append("  ").append(flag.toSummaryLine()).append("\n");
                if (useColors) {
                    sb.append(OutputPort.RESET);
                }
            }
            sb.append("\n");
        }

        // Summary
        sb.append("SUMMARY: ").append(report.summary()).append("\n");

        return sb.toString();
    }

    @Override
    public String formatHealthCheck(HealthCheckResult result) {
        StringBuilder sb = new StringBuilder();

        // Title
        if (useColors) {
            sb.append(OutputPort.BOLD).append(OutputPort.UNDERLINE);
        }
        sb.append("PORT ").append(result.port()).append(" HEALTH CHECK\n");
        if (useColors) {
            sb.append(OutputPort.RESET);
        }
        sb.append("======================\n");

        // TCP status
        String statusColor = result.isHealthy() ? OutputPort.GREEN : OutputPort.RED;
        if (useColors) {
            sb.append("TCP:     ").append(statusColor);
        } else {
            sb.append("TCP:     ");
        }
        sb.append(result.status().displayName().toUpperCase());
        sb.append(" (").append(result.responseTimeMs()).append("ms)");
        if (useColors) {
            sb.append(OutputPort.RESET);
        }
        sb.append("\n");

        // HTTP info if available
        result.httpInfoOpt().ifPresent(http -> {
            String httpColor = http.isSuccess() ? OutputPort.GREEN : OutputPort.YELLOW;
            if (useColors) {
                sb.append("HTTP:    ").append(httpColor);
            } else {
                sb.append("HTTP:    ");
            }
            sb.append(http.statusFormatted());
            sb.append(" (").append(http.responseTimeMs()).append("ms)");
            if (useColors) {
                sb.append(OutputPort.RESET);
            }
            sb.append("\n");

            if (http.contentType() != null) {
                sb.append("Content: ").append(http.contentType());
                sb.append(" (").append(http.contentLengthFormatted()).append(")");
                sb.append("\n");
            }
        });

        // SSL info if available
        result.sslInfoOpt().ifPresent(ssl -> {
            sb.append("\nSSL CERTIFICATE\n");
            sb.append("  Subject:  ").append(extractCN(ssl.subject())).append("\n");
            sb.append("  Issuer:   ").append(extractCN(ssl.issuer())).append("\n");
            sb.append("  Protocol: ").append(ssl.protocol()).append("\n");

            long daysLeft = ssl.daysUntilExpiry();
            String expiryColor = ssl.isExpired() ? OutputPort.RED :
                                ssl.expiresSoon() ? OutputPort.YELLOW : OutputPort.GREEN;

            if (useColors) {
                sb.append("  Expires:  ").append(expiryColor);
            } else {
                sb.append("  Expires:  ");
            }

            if (ssl.isExpired()) {
                sb.append("EXPIRED");
            } else if (daysLeft >= 0) {
                sb.append(daysLeft).append(" days");
            } else {
                sb.append("Unknown");
            }

            if (useColors) {
                sb.append(OutputPort.RESET);
            }
            sb.append("\n");
        });

        return sb.toString();
    }

    @Override
    public String mimeType() {
        return "text/plain";
    }

    @Override
    public String fileExtension() {
        return "txt";
    }

    private String colorize(String text, String color) {
        if (!useColors) return text;
        return color + text + OutputPort.RESET;
    }

    private String colorizeState(String state) {
        if (!useColors) return state;
        return switch (state) {
            case "LISTEN" -> OutputPort.GREEN + state + OutputPort.RESET;
            case "ESTABLISHED", "ESTAB" -> OutputPort.BLUE + state + OutputPort.RESET;
            case "TIME_WAIT", "CLOSE_WAIT" -> OutputPort.YELLOW + state + OutputPort.RESET;
            default -> state;
        };
    }

    private String extractCN(String dn) {
        // Extract CN from X.500 DN
        if (dn == null) return "";
        for (String part : dn.split(",")) {
            if (part.trim().toUpperCase().startsWith("CN=")) {
                return part.trim().substring(3);
            }
        }
        return dn.length() > 50 ? dn.substring(0, 47) + "..." : dn;
    }

    private String truncate(String text, int maxLength) {
        return config.truncate(text, maxLength);
    }
}

File: hefesto-core/src/feature/portinfo/output/table_formatter.rs
use crossterm::style::{Color, Stylize};

use hefesto_domain::config::HefestoConfig;
use hefesto_domain::portinfo::enriched_port_binding::EnrichedPortBinding;
use hefesto_domain::portinfo::health_check::HealthCheckResult;
use hefesto_domain::portinfo::network_overview::NetworkOverview;
use hefesto_domain::portinfo::port_binding::ConnectionState;
use hefesto_domain::portinfo::security::{SecurityReport, Severity};

use super::OutputFormatter;

/// Formats output as ASCII tables with optional crossterm colors.
pub struct TableFormatter {
    use_colors: bool,
}

impl TableFormatter {
    /// Creates a new `TableFormatter`.
    pub fn new(use_colors: bool) -> Self {
        Self { use_colors }
    }

    fn colorize(&self, text: &str, color: Color) -> String {
        if self.use_colors {
            text.with(color).to_string()
        } else {
            text.to_string()
        }
    }

    fn bold(&self, text: &str) -> String {
        if self.use_colors {
            text.bold().to_string()
        } else {
            text.to_string()
        }
    }

    fn underline(&self, text: &str) -> String {
        if self.use_colors {
            text.underlined().to_string()
        } else {
            text.to_string()
        }
    }

    fn colorize_state(&self, state: &ConnectionState) -> String {
        let text = state.as_str();
        if !self.use_colors {
            return text.to_string();
        }
        match state {
            ConnectionState::Listen => text.with(Color::Green).to_string(),
            ConnectionState::Established => text.with(Color::Blue).to_string(),
            ConnectionState::TimeWait | ConnectionState::CloseWait => {
                text.with(Color::Yellow).to_string()
            }
            _ => text.to_string(),
        }
    }

    fn extract_cn(dn: &str) -> &str {
        if dn.is_empty() {
            return "";
        }
        for part in dn.split(',') {
            let trimmed = part.trim();
            if trimmed.to_uppercase().starts_with("CN=") {
                return &trimmed[3..];
            }
        }
        if dn.len() > 50 {
            &dn[..47]
        } else {
            dn
        }
    }
}

impl Default for TableFormatter {
    fn default() -> Self {
        Self::new(true)
    }
}

impl OutputFormatter for TableFormatter {
    fn format(&self, bindings: &[EnrichedPortBinding]) -> String {
        if bindings.is_empty() {
            return "No port bindings found.".to_string();
        }

        let config = HefestoConfig::get();
        let max_process_width = config.max_process_name_width;
        let max_service_width = config.max_service_width;

        let proto_width = 5;
        let port_width = 6;
        let state_width = 12;
        let address_width = 15;
        let pid_width = 8;
        let mut process_width: usize = 15;
        let mut service_width: usize = 15;

        for b in bindings {
            let proc_len = b.process_name().len();
            let capped = if max_process_width > 0 {
                proc_len.min(max_process_width as usize)
            } else {
                proc_len
            };
            process_width = process_width.max(capped);

            if let Some(ref si) = b.service_info {
                let svc_len = si.name.len();
                let capped = if max_service_width > 0 {
                    svc_len.min(max_service_width as usize)
                } else {
                    svc_len
                };
                service_width = service_width.max(capped);
            }
        }

        let mut sb = String::new();

        // Header
        let header = format!(
            "{:<proto_w$}  {:<port_w$}  {:<state_w$}  {:<addr_w$}  {:<pid_w$}  {:<proc_w$}  {:<svc_w$}\n",
            "PROTO", "PORT", "STATE", "ADDRESS", "PID", "PROCESS", "SERVICE",
            proto_w = proto_width,
            port_w = port_width,
            state_w = state_width,
            addr_w = address_width,
            pid_w = pid_width,
            proc_w = process_width,
            svc_w = service_width,
        );
        let separator_len = header.len().saturating_sub(1);

        if self.use_colors {
            sb.push_str(&self.bold(&header));
        } else {
            sb.push_str(&header);
        }
        sb.push_str(&"-".repeat(separator_len));
        sb.push('\n');

        // Rows
        for b in bindings {
            let proto = b.protocol().as_str();
            let port = b.port().to_string();
            let state = b.state().as_str().to_string();
            let address = b.local_address().to_string();
            let pid = b.pid().to_string();
            let process = config.truncate_process_name(b.process_name());
            let service = b
                .service_info
                .as_ref()
                .map(|si| config.truncate_service(&si.name))
                .unwrap_or_else(|| "-".to_string());

            let proto_str = self.colorize(proto, Color::Cyan);
            let state_str = self.colorize_state(b.state());
            let process_str = self.colorize(&process, Color::Green);
            let address_str = if self.use_colors && b.is_exposed() {
                self.colorize(&address, Color::Yellow)
            } else {
                address.clone()
            };

            // For alignment, we format with the raw lengths then replace
            // with colored strings. Since ANSI codes add invisible chars,
            // we pad manually.
            sb.push_str(&format!(
                "{proto_str}{pad_proto}  {port:<port_w$}  {state_str}{pad_state}  {address_str}{pad_addr}  {pid:<pid_w$}  {process_str}{pad_proc}  {service}\n",
                pad_proto = " ".repeat(proto_width.saturating_sub(proto.len())),
                port_w = port_width,
                pad_state = " ".repeat(state_width.saturating_sub(state.len())),
                pad_addr = " ".repeat(address_width.saturating_sub(address.len())),
                pid_w = pid_width,
                pad_proc = " ".repeat(process_width.saturating_sub(process.len())),
            ));
        }

        sb
    }

    fn format_overview(&self, overview: &NetworkOverview) -> String {
        let mut sb = String::new();

        // Title
        let title = "NETWORK OVERVIEW\n";
        sb.push_str(&self.bold(&self.underline(title)));
        sb.push_str("================\n\n");

        // Statistics
        sb.push_str("STATISTICS\n");
        let stats = &overview.statistics;
        sb.push_str(&format!(
            "  Listening:     {} ports\n",
            stats.total_listening
        ));
        sb.push_str(&format!(
            "  Established:   {} connections\n",
            stats.total_established
        ));
        sb.push_str(&format!(
            "  TCP: {} | UDP: {}\n",
            stats.tcp_count, stats.udp_count
        ));
        sb.push_str(&format!(
            "  Exposed (0.0.0.0): {} | Local (127.0.0.1): {}\n\n",
            stats.exposed_count, stats.local_only_count
        ));

        // Exposed ports
        let exposed = overview.exposed_ports();
        if !exposed.is_empty() {
            sb.push_str("EXPOSED PORTS (Network Accessible)\n");
            for b in &exposed {
                let service_tag = b
                    .service_info
                    .as_ref()
                    .map(|si| si.to_tag())
                    .unwrap_or_default();
                let line = format!(
                    "  {} :{:<6} {:<12} {:<15} pid={}\n",
                    b.protocol(),
                    b.port(),
                    b.process_name(),
                    service_tag,
                    b.pid()
                );
                if self.use_colors {
                    sb.push_str(&self.colorize(&line, Color::Yellow));
                } else {
                    sb.push_str(&line);
                }
            }
            sb.push('\n');
        }

        // Group by process
        let by_pid = overview.by_pid();
        if !by_pid.is_empty() {
            sb.push_str("PROCESSES\n");
            let mut pids: Vec<u32> = by_pid.keys().copied().collect();
            pids.sort();

            for pid in pids {
                if let Some(ports) = by_pid.get(&pid) {
                    if ports.is_empty() {
                        continue;
                    }
                    let first = ports[0];

                    let mut header_line =
                        format!("  {} (pid {})", first.process_name(), first.pid());
                    if !first.user().is_empty() {
                        header_line.push_str(&format!(" - {}", first.user()));
                    }

                    sb.push_str(&self.colorize(&header_line, Color::Green));
                    sb.push('\n');

                    for b in ports {
                        let service_tag = b
                            .service_info
                            .as_ref()
                            .map(|si| si.to_tag())
                            .unwrap_or_default();
                        let exposure = if b.is_exposed() {
                            "0.0.0.0"
                        } else if b.is_local_only() {
                            "127.0.0.1"
                        } else {
                            b.local_address()
                        };
                        sb.push_str(&format!(
                            "    :{}  {}  {}  {}\n",
                            b.port(),
                            b.state(),
                            exposure,
                            service_tag
                        ));
                    }
                }
            }
        }

        sb
    }

    fn format_security_report(&self, report: &SecurityReport) -> String {
        let mut sb = String::new();

        // Title
        let title = "SECURITY ANALYSIS\n";
        sb.push_str(&self.bold(&self.underline(title)));
        sb.push_str("=================\n\n");

        if report.is_clean() {
            sb.push_str(&self.colorize("No security issues found.\n", Color::Green));
            return sb;
        }

        // Group by severity
        let severity_order = [
            Severity::Critical,
            Severity::High,
            Severity::Warning,
            Severity::Info,
        ];

        for severity in &severity_order {
            let flags: Vec<_> = report
                .findings
                .iter()
                .filter(|f| &f.severity == severity)
                .collect();
            if flags.is_empty() {
                continue;
            }

            let color = match severity {
                Severity::Critical | Severity::High => Color::Red,
                Severity::Warning => Color::Yellow,
                Severity::Info => Color::Cyan,
            };

            let header = format!(
                "{} ({})\n",
                severity.display_name().to_uppercase(),
                flags.len()
            );
            sb.push_str(&self.colorize(&self.bold(&header), color));

            for flag in &flags {
                let line = format!("  {}\n", flag.to_summary_line());
                sb.push_str(&self.colorize(&line, color));
            }
            sb.push('\n');
        }

        // Summary
        sb.push_str(&format!("SUMMARY: {}\n", report.summary()));

        sb
    }

    fn format_health_check(&self, result: &HealthCheckResult) -> String {
        let mut sb = String::new();

        // Title
        let title = format!("PORT {} HEALTH CHECK\n", result.port);
        sb.push_str(&self.bold(&self.underline(&title)));
        sb.push_str("======================\n");

        // TCP status
        let status_color = if result.is_healthy() {
            Color::Green
        } else {
            Color::Red
        };

        let status_text = format!(
            "{} ({}ms)",
            result.status.display_name().to_uppercase(),
            result.response_time_ms
        );
        sb.push_str("TCP:     ");
        sb.push_str(&self.colorize(&status_text, status_color));
        sb.push('\n');

        // HTTP info if available
        if let Some(ref http) = result.http_info {
            let http_color = if http.is_success() {
                Color::Green
            } else {
                Color::Yellow
            };

            let http_text = format!("{} ({}ms)", http.status_formatted(), http.response_time_ms);
            sb.push_str("HTTP:    ");
            sb.push_str(&self.colorize(&http_text, http_color));
            sb.push('\n');

            if !http.content_type.is_empty() {
                sb.push_str(&format!(
                    "Content: {} ({})\n",
                    http.content_type,
                    http.content_length_formatted()
                ));
            }
        }

        // SSL info if available
        if let Some(ref ssl) = result.ssl_info {
            sb.push_str("\nSSL CERTIFICATE\n");
            sb.push_str(&format!("  Subject:  {}\n", Self::extract_cn(&ssl.subject)));
            sb.push_str(&format!("  Issuer:   {}\n", Self::extract_cn(&ssl.issuer)));
            sb.push_str(&format!("  Protocol: {}\n", ssl.protocol));

            let days_left = ssl.days_until_expiry();
            let expiry_color = if ssl.is_expired() {
                Color::Red
            } else if ssl.expires_soon() {
                Color::Yellow
            } else {
                Color::Green
            };

            let expiry_text = if ssl.is_expired() {
                "EXPIRED".to_string()
            } else if days_left >= 0 {
                format!("{days_left} days")
            } else {
                "Unknown".to_string()
            };

            sb.push_str("  Expires:  ");
            sb.push_str(&self.colorize(&expiry_text, expiry_color));
            sb.push('\n');
        }

        sb
    }

    fn mime_type(&self) -> &str {
        "text/plain"
    }

    fn file_extension(&self) -> &str {
        "txt"
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use hefesto_domain::portinfo::port_binding::{PortBinding, Protocol};

    #[test]
    fn test_format_empty() {
        let formatter = TableFormatter::new(false);
        let result = formatter.format(&[]);
        assert_eq!(result, "No port bindings found.");
    }

    #[test]
    fn test_format_single_binding() {
        let formatter = TableFormatter::new(false);
        let binding = PortBinding::listen(8080, Protocol::Tcp, 1234, "java");
        let enriched = EnrichedPortBinding::from_binding(binding);
        let result = formatter.format(&[enriched]);
        assert!(result.contains("PROTO"));
        assert!(result.contains("8080"));
        assert!(result.contains("java"));
        assert!(result.contains("LISTEN"));
    }

    #[test]
    fn test_extract_cn() {
        assert_eq!(
            TableFormatter::extract_cn("CN=example.com,O=Org"),
            "example.com"
        );
        assert_eq!(TableFormatter::extract_cn(""), "");
    }

    #[test]
    fn test_format_health_check() {
        use hefesto_domain::portinfo::health_check::HealthStatus;

        let formatter = TableFormatter::new(false);
        let result = HealthCheckResult::tcp(8080, HealthStatus::Reachable, 42, "OK");
        let output = formatter.format_health_check(&result);
        assert!(output.contains("8080"));
        assert!(output.contains("REACHABLE"));
        assert!(output.contains("42ms"));
    }

    #[test]
    fn test_format_security_clean() {
        use hefesto_domain::portinfo::security::SecurityReport;

        let formatter = TableFormatter::new(false);
        let report = SecurityReport::empty();
        let output = formatter.format_security_report(&report);
        assert!(output.contains("No security issues found"));
    }
}
